File: apps/web/services/courses/quizFlags.ts
import { getAPIUrl } from '@services/config/config'
import {
  RequestBodyWithAuthHeader,
  errorHandling,
  getResponseMetadata,
} from '@services/utils/ts/requests'

export type QuizFlagReason = 'incorrect_answer' | 'unclear_wording' | 'typo' | 'other'
export type QuizFlagStatus = 'open' | 'resolved' | 'dismissed'

export interface QuizFlagAuthor {
  id: number
  user_uuid: string
  username: string
  first_name: string
  last_name: string
}

export interface QuizFlag {
  flag_uuid: string
  activity_uuid: string
  activity_name: string
  quiz_id: string
  question_id: string
  question_text_snapshot: string
  reason: QuizFlagReason
  note: string | null
  status: QuizFlagStatus
  flagged_by: QuizFlagAuthor | null
  resolved_by: QuizFlagAuthor | null
  resolved_at: string | null
  creation_date: string
}

export async function createQuizFlag(
  activity_uuid: string,
  quiz_id: string,
  question_id: string,
  reason: QuizFlagReason,
  note: string | null,
  access_token: string | null | undefined
): Promise<QuizFlag> {
  const result: any = await fetch(
    `${getAPIUrl()}activities/${activity_uuid}/quiz_flags`,
    RequestBodyWithAuthHeader(
      'POST',
      { quiz_id, question_id, reason, note },
      null,
      access_token || undefined
    )
  )
  const res = await getResponseMetadata(result)
  if (!res.success) {
    const detail = res.data?.detail || res.data?.message || res.data
    const message = typeof detail === 'string' ? detail : JSON.stringify(detail)
    const error: any = new Error(message)
    error.status = res.status
    error.detail = detail
    throw error
  }
  return res.data
}

export async function getCourseQuizFlags(
  course_uuid: string,
  status: QuizFlagStatus | null,
  access_token: string | null | undefined
): Promise<QuizFlag[]> {
  const url = `${getAPIUrl()}courses/${course_uuid}/quiz_flags${status ? `?status=${status}` : ''}`
  const result: any = await fetch(
    url,
    RequestBodyWithAuthHeader('GET', null, null, access_token || undefined)
  )
  const res = await errorHandling(result)
  return res
}

export async function resolveQuizFlag(
  flag_uuid: string,
  status: QuizFlagStatus,
  access_token: string | null | undefined
): Promise<QuizFlag> {
  const result: any = await fetch(
    `${getAPIUrl()}quiz_flags/${flag_uuid}`,
    RequestBodyWithAuthHeader('PUT', { status }, null, access_token || undefined)
  )
  const res = await errorHandling(result)
  return res
}

File: apps/api/src/db/courses/quiz_flags.py
"""Student-flaggable quiz questions.

A quiz question is not its own database row (it lives inside a `blockQuiz`
node in an Activity's Prosemirror `content` document — see
services/ai/quiz.py's module docstring), so a flag cannot be a foreign key
to a question row. Instead it references the (activity, quiz, question) by
id and keeps a text SNAPSHOT of the question at flag time — content can be
edited (or the question deleted) after a flag is raised, and the review
queue must still show the instructor what a student actually saw, not
whatever the block currently contains.

Scope: only the ungraded, in-content `blockQuiz` self-check block. The
separate graded ASSIGNMENT quiz task type has a different question shape
entirely (see services/ai/quiz.py) and is not covered here.
"""

from typing import Optional
from enum import Enum
from sqlalchemy import Column, ForeignKey, Integer, String, Text, Index
from sqlmodel import Field, SQLModel


class QuizFlagReason(str, Enum):
    INCORRECT_ANSWER = "incorrect_answer"
    UNCLEAR_WORDING = "unclear_wording"
    TYPO = "typo"
    OTHER = "other"


class QuizFlagStatus(str, Enum):
    OPEN = "open"
    RESOLVED = "resolved"
    DISMISSED = "dismissed"


class QuizQuestionFlag(SQLModel, table=True):
    __tablename__ = "quiz_question_flag"
    __table_args__ = (
        Index("ix_quiz_question_flag_flag_uuid", "flag_uuid"),
        Index("ix_quiz_question_flag_course_status", "course_id", "status"),
        Index("ix_quiz_question_flag_activity_question", "activity_id", "question_id"),
        {"extend_existing": True},
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    flag_uuid: str = Field(default="", max_length=100)
    org_id: int = Field(
        sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), nullable=False)
    )
    course_id: int = Field(
        sa_column=Column(Integer, ForeignKey("course.id", ondelete="CASCADE"), nullable=False)
    )
    activity_id: int = Field(
        sa_column=Column(Integer, ForeignKey("activity.id", ondelete="CASCADE"), nullable=False)
    )
    quiz_id: str = Field(sa_column=Column(String(100), nullable=False))
    question_id: str = Field(sa_column=Column(String(100), nullable=False))
    question_text_snapshot: str = Field(default="", sa_column=Column(Text, nullable=False))
    reason: QuizFlagReason = Field(sa_column=Column(String(30), nullable=False))
    note: Optional[str] = Field(default=None, sa_column=Column(Text, nullable=True))
    status: QuizFlagStatus = Field(
        default=QuizFlagStatus.OPEN, sa_column=Column(String(20), nullable=False, server_default="open")
    )
    # SET NULL rather than CASCADE: a flag stays useful to instructors as a
    # content-quality signal even if the student who raised it, or the
    # instructor who resolved it, later leaves the org / is deleted.
    flagged_by_user_id: Optional[int] = Field(
        default=None, sa_column=Column(Integer, ForeignKey("user.id", ondelete="SET NULL"), nullable=True)
    )
    resolved_by_user_id: Optional[int] = Field(
        default=None, sa_column=Column(Integer, ForeignKey("user.id", ondelete="SET NULL"), nullable=True)
    )
    resolved_at: Optional[str] = None
    creation_date: str = ""
    update_date: str = ""


# ---------------------------------------------------------------------------
# Pydantic schemas
# ---------------------------------------------------------------------------
from pydantic import BaseModel


class QuizFlagCreate(BaseModel):
    quiz_id: str
    question_id: str
    reason: QuizFlagReason
    note: Optional[str] = None


class QuizFlagAuthor(BaseModel):
    id: int
    user_uuid: str
    username: str
    first_name: str
    last_name: str


class QuizFlagRead(BaseModel):
    flag_uuid: str
    activity_uuid: str
    activity_name: str
    quiz_id: str
    question_id: str
    question_text_snapshot: str
    reason: QuizFlagReason
    note: Optional[str] = None
    status: QuizFlagStatus
    flagged_by: Optional[QuizFlagAuthor] = None
    resolved_by: Optional[QuizFlagAuthor] = None
    resolved_at: Optional[str] = None
    creation_date: str


class QuizFlagResolve(BaseModel):
    status: QuizFlagStatus
